packages/taiko-client/driver/chain_syncer/blob/soft_block.go

```diff
@@ -0,0 +1,291 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to update fork choice: %w", err)
+	}
+	if fcRes.PayloadStatus.Status != engine.VALID {
+		return nil, fmt.Errorf("unexpected ForkchoiceUpdate response status: %s", fcRes.PayloadStatus.Status)
+	}
+	if fcRes.PayloadID == nil {
+		return nil, errors.New("empty payload ID")
+	}
+
+	// Step 2, get the payload
+	payload, err := s.rpc.L2Engine.GetPayload(ctx, fcRes.PayloadID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get payload: %w", err)
+	}
+
+	log.Info(
+		"Soft block payload",
+		"blockID", blockID,
+		"baseFee", utils.WeiToGWei(payload.BaseFeePerGas),
+		"number", payload.Number,
+		"hash", payload.BlockHash,
+		"gasLimit", payload.GasLimit,
+		"gasUsed", payload.GasUsed,
+		"timestamp", payload.Timestamp,
+		"withdrawalsHash", payload.WithdrawalsHash,
+	)
+
+	// Step 3, execute the payload
+	execStatus, err := s.rpc.L2Engine.NewPayload(ctx, payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new payload: %w", err)
+	}
+	if execStatus.Status != engine.VALID {
+		return nil, fmt.Errorf("unexpected NewPayload response status: %s", execStatus.Status)
+	}
+
+	lastVerifiedBlockHash, err := s.rpc.GetLastVerifiedBlockHash(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch last verified block hash: %w", err)
+	}
+
+	canonicalHead, err := s.rpc.L2.HeadL1Origin(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch canonical head: %w", err)
+	}
+
+	// Step 4, update the fork choice
+	fc = &engine.ForkchoiceStateV1{
+		HeadBlockHash:      payload.BlockHash,
+		SafeBlockHash:      canonicalHead.L2BlockHash,
+		FinalizedBlockHash: lastVerifiedBlockHash,
+	}
+	fcRes, err = s.rpc.L2Engine.ForkchoiceUpdate(ctx, fc, nil)
+	if err != nil {
+		return nil, err
+	}
+	if fcRes.PayloadStatus.Status != engine.VALID {
+		return nil, fmt.Errorf("unexpected ForkchoiceUpdate response status: %s", fcRes.PayloadStatus.Status)
+	}
+
+	header, err := s.rpc.L2.HeaderByHash(ctx, payload.BlockHash)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info(
+		"⏰ New soft L2 block inserted",
+		"blockID", blockID,
+		"hash", header.Hash(),
+		"transactions", len(payload.Transactions),
+		"baseFee", utils.WeiToGWei(header.BaseFee),
+		"withdrawals", len(payload.Withdrawals),
+		"endOfBlock", attributes.L1Origin.EndOfBlock,
+		"endOfPreconf", attributes.L1Origin.EndOfPreconf,
+	)
+
+	return header, nil
+}
+
+// RemoveSoftBlocks removes soft blocks from the L2 execution engine's blockchain.
+func (s *Syncer) RemoveSoftBlocks(ctx context.Context, newLastBlockID uint64) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	newHead, err := s.rpc.L2.HeaderByNumber(ctx, new(big.Int).SetUint64(newLastBlockID))
+	if err != nil {
+		return err
+	}
+
+	lastVerifiedBlockHash, err := s.rpc.GetLastVerifiedBlockHash(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch last verified block hash: %w", err)
+	}
+
+	fc := &engine.ForkchoiceStateV1{
```

Comment:
We can only remove `softblocks`, so the `SafeBlockHash` and `FinalizedBlockHash` aren't changed and there is no need to set them. Besides the `SafeBlockHash` is not right, it should be equal to `canonicalHeadL1Origin.L2BlockHash`.